service/src/main/java/crawlercommons/urlfrontier/service/cluster/Hearbeat.java

```diff
@@ -51,9 +54,18 @@ public void setListener(HeartbeatListener listener) {
         this.listener = listener;
     }
 
+    public void setService(AbstractFrontierService service) {
+        this.service = service;
+    }
+
     @Override
     public void run() {
 
+        // wait until the service is set up to prevent a race condition
+        if (service != null) {
```

Comment:
Ideally I would have preferred to discuss this in a separate PR and be able to go ahead and merge the rest.
Shouldn't the call to _setService_ be used by other implementations as well?
Having _getHostAndPort()_  set could not be the only indication that the service is ready - maybe we'd need something more generic like _isReady()_ and have that logic implemented in each service?

Reply:
I will put this modification in a separate PR then. Also I will create another one for issue #73 to have one PR per issue (makes totally sense).
Yes, a more generic method `isReady()` could substitute the querying on `service.getHostAndPort() == null`. I will add it this way